File: src/spektrafilm/model/couplers.py
import numpy as np
from opt_einsum import contract

from spektrafilm.runtime.params_schema import DirCouplersParams
from spektrafilm.utils.fast_gaussian_filter import fast_gaussian_filter
from spektrafilm.model.density_curves import interpolate_exposure_to_density

def compute_density_curves_before_dir_couplers(density_curves, log_exposure, dir_couplers_matrix, positive=False):
    """
    DIR couplers affect the same layer by increasing contrast.
    I suppose that in the design of a film this is taken into account, and the final film has well behaved density curves.
    In order to get final curves for gray ramps equal to the input data, the density curves before the effect of the couplers are needed.

    Args:
        density_curves (numpy.array): Characteristic density curves of the film after the application of DIR couplers
        log_exposure (numpy.array): The image as log_exposure
        dir_couplers_matrix (_type_): DIR couplers matrix computed with compute_dir_couplers_matrix()

    Returns:
        numpy.array: Corrected density curves before the effect of DIR couplers
    """
    
    if positive:
        # We are asusming that interimage effects in positve film are acting in the silver development stage
        # We are also assuming that silver density is d_max - d
        density_curves_silver = np.nanmax(density_curves, axis=0) - density_curves
    else:
        density_curves_silver = np.copy(density_curves)
    
    couplers_amount_curves = contract('jk, km->jm', density_curves_silver, dir_couplers_matrix)
    log_exposure_0 = log_exposure[:,None] - couplers_amount_curves
    density_curves_corrected = np.zeros_like(density_curves)
    for i in np.arange(3):
        if positive:
            density_curves_corrected[:,i] = -np.interp(log_exposure, log_exposure_0[:,i], -density_curves[:,i])
        else:
            density_curves_corrected[:,i] = np.interp(log_exposure, log_exposure_0[:,i], density_curves[:,i])
    return density_curves_corrected


def compute_dir_couplers_matrix(couplers_params: DirCouplersParams = DirCouplersParams()):
    """
    Compute the inhibitors matrix using a simple diffusion model across layers.

    Parameters:
    amount_rgb (list of float): Amounts of dir couplers for RGB channels. Default is [0.7,0.7,0.5]. Typically 0-1 range.
    layer_diffusion (float): Sigma for gaussian diffusion distance of dir couplers. Default is 1.

    Returns:
    numpy.ndarray: The computed inhibitors matrix.
    Row index is the donor/source layer that releases inhibitor.
    Column index is the receiving/affected layer whose exposure is reduced.
    """
    
    M_self = np.array(couplers_params.gamma_samelayer_rgb)*couplers_params.inhibition_samelayer
    M_self = np.diag(M_self)
    M_inter = np.zeros((3,3))
    # Off-diagonal terms follow the same convention: donor row, receiver column.
    M_inter[0,1] = couplers_params.gamma_interlayer_r_to_gb[0]
    M_inter[0,2] = couplers_params.gamma_interlayer_r_to_gb[1]
    M_inter[1,0] = couplers_params.gamma_interlayer_g_to_rb[0]
    M_inter[1,2] = couplers_params.gamma_interlayer_g_to_rb[1]
    M_inter[2,0] = couplers_params.gamma_interlayer_b_to_rg[0]
    M_inter[2,1] = couplers_params.gamma_interlayer_b_to_rg[1]
    M_inter *= couplers_params.inhibition_interlayer
    return M_self + M_inter

def compute_exposure_correction_dir_couplers(log_raw, density_cmy, density_max,
                                             dir_couplers_matrix, diffusion_size_pixel,
                                             high_exposure_couplers_shift=0.0,
                                             positive=False):
    """
    Apply coupler inhibitors to the raw data based on density curves and inhibitor values.
    Coupler inhibitors are released when density is formed in the emulsion layers.
    If a layer is dense, the inhibitors are released to prevent further density formation in neighboring layers.
    Also self-inhibitors in the same layer, after spatial diffusion, can prevent further density formation
    in nearby areas, adding a local contrast effect.

    Parameters:
    raw (numpy.ndarray): The raw data to which inhibitors will be applied.
    density_cmy (numpy.ndarray): The density values for each layer.
    density_max (float): The maximum density value achievable for each layer, used for normalization.
    dir_couplers_matrix (numpy.ndarray): The inhibitors matrix. Fisrt index is the input layer, second index is the output layer.
    diffusion_size_pixel (int): The size of the gaussian filter for the diffusion of the inhibitors in xy.
    high_exposure_couplers_shift (float): if overexposure increases saturation, this will increase the inhibitors effect at higher density
    
    Returns:
    numpy.ndarray: The modified raw exposure data after applying the effect of inhibitors.
    """
    if positive:
        density_silver = density_max - density_cmy
    else:
        density_silver = np.copy(density_cmy)
    density_silver += high_exposure_couplers_shift*density_silver**2
    # density_silver[..., k] generated in donor layer k contributes to receiver m
    # through dir_couplers_matrix[k, m].
    log_raw_correction = contract('ijk, km->ijm', density_silver, dir_couplers_matrix)
    if diffusion_size_pixel>0:
        # log_raw_correction = gaussian_filter(log_raw_correction, (diffusion_size_pixel, diffusion_size_pixel, 0))
        log_raw_correction = fast_gaussian_filter(log_raw_correction, diffusion_size_pixel)
    log_raw_corrected = log_raw - log_raw_correction
    return log_raw_corrected


def apply_density_correction_dir_couplers(
    density_cmy,
    log_raw,
    pixel_size_um,
    log_exposure,
    density_curves,
    dir_couplers,
    profile_type,
    gamma_factor=1.0,
):
    if not dir_couplers.active:
        return density_cmy

    positive = profile_type == 'positive'
    
    couplers_matrix = compute_dir_couplers_matrix(dir_couplers)
    couplers_matrix *= dir_couplers.amount
    
    density_curves_0 = compute_density_curves_before_dir_couplers(
        density_curves,
        log_exposure,
        couplers_matrix,
        positive=positive,
    )
    density_max = np.nanmax(density_curves, axis=0)
    diffusion_size_pixel = dir_couplers.diffusion_size_um / pixel_size_um
    log_raw_0 = compute_exposure_correction_dir_couplers(
        log_raw,
        density_cmy,
        density_max,
        couplers_matrix,
        diffusion_size_pixel,
        positive=positive,
    )
    return interpolate_exposure_to_density(log_raw_0, density_curves_0, log_exposure, gamma_factor)

if __name__=='__main__':
    # # Test the raw correction coupler inhibitors
    # log_raw = np.ones((4,4,3))
    # density_cmy = np.ones((4,4,3))
    # density_max = 2.2
    # couplers_amount = [0.9,0.7,0.5]
    # diffusion_size_pixel = 2
    # log_raw = raw_correction_dir_couplers(log_raw, density_cmy, density_max, couplers_amount, diffusion_size_pixel)
    # print(log_raw)

    couplers_params = DirCouplersParams()
    M = compute_dir_couplers_matrix(couplers_params)
    print(M)

File: spectralfilm-rs/spectralfilm-core/src/model/couplers.rs
//! DIR (developer inhibitor-releasing) coupler model.
//!
//! Mirrors `spektrafilm/model/couplers.py`.

use crate::model::density_curves::interpolate_exposure_to_density;
use crate::utils::gaussian::gaussian_filter_2d;

// ---------------------------------------------------------------------------
// Parameters
// ---------------------------------------------------------------------------

#[derive(Debug, Clone)]
pub struct DirCouplersParams {
    pub active: bool,
    pub amount: f64,
    pub inhibition_samelayer: f64,
    pub inhibition_interlayer: f64,
    /// Same-layer gamma for R, G, B channels.
    pub gamma_samelayer_rgb: [f64; 3],
    /// Interlayer R→G, R→B.
    pub gamma_interlayer_r_to_gb: [f64; 2],
    /// Interlayer G→R, G→B.
    pub gamma_interlayer_g_to_rb: [f64; 2],
    /// Interlayer B→R, B→G.
    pub gamma_interlayer_b_to_rg: [f64; 2],
    /// Diffusion sigma (µm) of the inhibitor cloud.
    pub diffusion_size_um: f64,
}

impl Default for DirCouplersParams {
    fn default() -> Self {
        Self {
            active: true,
            amount: 1.0,
            inhibition_samelayer: 1.0,
            inhibition_interlayer: 1.0,
            gamma_samelayer_rgb: [0.341, 0.324, 0.273],
            gamma_interlayer_r_to_gb: [0.355, 0.305],
            gamma_interlayer_g_to_rb: [0.154, 0.358],
            gamma_interlayer_b_to_rg: [0.171, 0.225],
            diffusion_size_um: 20.0,
        }
    }
}

// ---------------------------------------------------------------------------
// 3×3 coupling matrix
// ---------------------------------------------------------------------------

/// Build the 3×3 DIR coupler matrix M where M[donor, receiver] = gamma.
/// Row = donor layer, column = affected layer.
pub fn compute_dir_couplers_matrix(p: &DirCouplersParams) -> [[f64; 3]; 3] {
    let mut m = [[0.0f64; 3]; 3];
    // Diagonal (same-layer)
    m[0][0] = p.gamma_samelayer_rgb[0] * p.inhibition_samelayer;
    m[1][1] = p.gamma_samelayer_rgb[1] * p.inhibition_samelayer;
    m[2][2] = p.gamma_samelayer_rgb[2] * p.inhibition_samelayer;
    // Off-diagonal (interlayer)
    m[0][1] = p.gamma_interlayer_r_to_gb[0] * p.inhibition_interlayer;
    m[0][2] = p.gamma_interlayer_r_to_gb[1] * p.inhibition_interlayer;
    m[1][0] = p.gamma_interlayer_g_to_rb[0] * p.inhibition_interlayer;
    m[1][2] = p.gamma_interlayer_g_to_rb[1] * p.inhibition_interlayer;
    m[2][0] = p.gamma_interlayer_b_to_rg[0] * p.inhibition_interlayer;
    m[2][1] = p.gamma_interlayer_b_to_rg[1] * p.inhibition_interlayer;
    m
}

// ---------------------------------------------------------------------------
// Density curves before DIR couplers (invert coupler effect from published curves)
// ---------------------------------------------------------------------------

pub fn compute_density_curves_before_dir_couplers(
    density_curves: &[f64],  // N_LE × 3
    log_exposure: &[f64],    // N_LE
    couplers_matrix: &[[f64; 3]; 3],
    positive: bool,
    n_le: usize,
) -> Vec<f64> {
    let mut dc_silver = density_curves.to_vec();
    if positive {
        // silver density = d_max - d
        let mut dmax = [0.0f64; 3];
        for le in 0..n_le {
            for ch in 0..3 {
                let v = density_curves[le * 3 + ch];
                if v.is_finite() && v > dmax[ch] { dmax[ch] = v; }
            }
        }
        for le in 0..n_le {
            for ch in 0..3 {
                dc_silver[le * 3 + ch] = dmax[ch] - density_curves[le * 3 + ch];
            }
        }
    }

    // coupler_amount[le, receiver] = sum_donor( dc_silver[le,donor] * M[donor,receiver] )
    let mut coupler_amount = vec![0.0f64; n_le * 3];
    for le in 0..n_le {
        for receiver in 0..3 {
            let mut sum = 0.0;
            for donor in 0..3 {
                sum += dc_silver[le * 3 + donor] * couplers_matrix[donor][receiver];
            }
            coupler_amount[le * 3 + receiver] = sum;
        }
    }

    // Shifted log-exposure axis: le_0[le, ch] = log_exposure[le] - coupler_amount[le,ch]
    let mut dc_corrected = vec![0.0f64; n_le * 3];
    for ch in 0..3 {
        // Build shifted x and original y, then re-interpolate
        let le_shifted: Vec<f64> = (0..n_le).map(|k| log_exposure[k] - coupler_amount[k * 3 + ch]).collect();
        let y_orig: Vec<f64> = (0..n_le).map(|k| density_curves[k * 3 + ch]).collect();
        for le in 0..n_le {
            let x = log_exposure[le];
            let v = if positive {
                -interp1d_sorted_clamped(x, &le_shifted, &y_orig.iter().map(|&v| -v).collect::<Vec<_>>())
            } else {
                interp1d_sorted_clamped(x, &le_shifted, &y_orig)
            };
            dc_corrected[le * 3 + ch] = v;
        }
    }
    dc_corrected
}

fn interp1d_sorted_clamped(x: f64, xs: &[f64], ys: &[f64]) -> f64 {
    use crate::model::density_curves::interp1d_clamped;
    let n = xs.len();
    interp1d_clamped(x, n, |k| xs[k], |k| ys[k])
}

// ---------------------------------------------------------------------------
// Exposure correction from dir couplers
// ---------------------------------------------------------------------------

pub fn compute_exposure_correction_dir_couplers(
    log_raw: &[f64],          // N_PX × 3
    density_cmy: &[f64],      // N_PX × 3
    density_max: &[f64; 3],
    couplers_matrix: &[[f64; 3]; 3],
    diffusion_size_px: f64,
    positive: bool,
    width: usize,
    height: usize,
) -> Vec<f64> {
    let n_px = width * height;

    // Silver density
    let mut density_silver: Vec<f64> = density_cmy.to_vec();
    if positive {
        for px in 0..n_px {
            for ch in 0..3 {
                density_silver[px * 3 + ch] = density_max[ch] - density_cmy[px * 3 + ch];
            }
        }
    }

    // log_raw_correction[px, receiver] = sum_donor( density_silver[px,donor] * M[donor,receiver] )
    let mut correction = vec![0.0f64; n_px * 3];
    for px in 0..n_px {
        for receiver in 0..3 {
            let mut sum = 0.0;
            for donor in 0..3 {
                sum += density_silver[px * 3 + donor] * couplers_matrix[donor][receiver];
            }
            correction[px * 3 + receiver] = sum;
        }
    }

    // Spatial diffusion
    if diffusion_size_px > 0.0 {
        for ch in 0..3 {
            let mut slice: Vec<f64> = (0..n_px).map(|px| correction[px * 3 + ch]).collect();
            slice = gaussian_filter_2d(&slice, width, height, diffusion_size_px);
            for px in 0..n_px { correction[px * 3 + ch] = slice[px]; }
        }
    }

    // Apply correction
    let mut out = vec![0.0f64; n_px * 3];
    for px in 0..n_px {
        for ch in 0..3 {
            out[px * 3 + ch] = log_raw[px * 3 + ch] - correction[px * 3 + ch];
        }
    }
    out
}

// ---------------------------------------------------------------------------
// Public apply function
// ---------------------------------------------------------------------------

pub fn apply_density_correction_dir_couplers(
    density_cmy: &[f64],
    log_raw: &[f64],
    pixel_size_um: f64,
    log_exposure: &[f64],
    density_curves: &[f64],
    p: &DirCouplersParams,
    profile_type: &str,
    gamma_factor: f64,
    shape: (usize, usize),
) -> Vec<f64> {
    if !p.active {
        return density_cmy.to_vec();
    }
    let positive = profile_type == "positive";
    let (h, w) = shape;
    let n_le = log_exposure.len();

    let mut m = compute_dir_couplers_matrix(p);
    for row in &mut m { for v in row.iter_mut() { *v *= p.amount; } }

    let dc_before = compute_density_curves_before_dir_couplers(
        density_curves, log_exposure, &m, positive, n_le,
    );

    let mut dmax = [0.0f64; 3];
    for le in 0..n_le {
        for ch in 0..3 {
            let v = density_curves[le * 3 + ch];
            if v.is_finite() && v > dmax[ch] { dmax[ch] = v; }
        }
    }

    let diffusion_px = p.diffusion_size_um / pixel_size_um;
    let log_raw_0 = compute_exposure_correction_dir_couplers(
        log_raw, density_cmy, &dmax, &m, diffusion_px, positive, w, h,
    );

    interpolate_exposure_to_density(&log_raw_0, log_exposure, &dc_before, [gamma_factor; 3])
}
